dissect/target/plugins/apps/edr/velociraptor.py

```diff
@@ -0,0 +1,98 @@
+import json
+import re
+import urllib
+from functools import lru_cache
+from typing import Iterator
+
+from flow.record import RecordDescriptor
+
+from dissect.target.exceptions import UnsupportedPluginError
+from dissect.target.helpers.record import DynamicDescriptor, TargetRecordDescriptor
+from dissect.target.plugin import Plugin, export
+from dissect.target.target import Target
+
+VELOCIRAPTOR_RESULTS = "/$velociraptor_results$"
+ISO_8601_PATTERN = r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})?"
+
+
+class VelociraptorRecordBuilder:
+    def __init__(self, os_type: str, artifact_name: str):
+        self._create_event_descriptor = lru_cache(4096)(self._create_event_descriptor)
+        self.RECORD_NAME = f"velociraptor/{os_type}/{artifact_name}"
+
+    def read_record(self, object: dict, target: Target):
+        """Builds a Velociraptor record"""
+        record_values = {}
+        record_fields = []
+
+        record_values["_target"] = target
+
+        for key, value in object.items():
+            # Reserved by flow.record
+            if key.startswith("_"):
+                continue
+
+            key = key.lower()
+
+            if re.match(ISO_8601_PATTERN, str(value)):
+                record_type = "datetime"
+            elif isinstance(value, list):
+                record_type = "string[]"
+            elif isinstance(value, int):
+                record_type = "varint"
+            elif key == "hash":
+                record_type = "digest"
+                value = (value["md5"] or None, value["sha1"] or None, value["sha256"] or None)
+            elif isinstance(value, str):
+                record_type = "string"
+            # FIXME: Why is there no record type dict?
```

Comment:
We have nested records, if that works for you. Otherwise you can raise an issue in `flow.record`.

Reply:
How do nested records work?

Reply:
There exists a `record` field type and the value is just any record. You can also use `record[]` for lists of records.

Reply:
https://github.com/fox-it/flow.record/blob/f4eba70f65e0f9a268f5e1421b06df1612a3acee/tests/test_packer.py#L200

Apparently we have `dictlist` too I just noticed 😄.

Reply:
Thanks, `record` works fine